apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,483 @@
+            need_joinids
+            and column_names is not None
+            and "soma_joinid" not in column_names
+        ):
+            query_columns = ["soma_joinid"] + list(column_names)
+
+        arrow_table = axis_df.read(
+            ids=axis_query.coords,
+            value_filter=axis_query.value_filter,
+            column_names=query_columns,
+        ).concat()
+
+        if need_joinids:
+            self._joinids[axis] = arrow_table.column("soma_joinid").combine_chunks()
+        assert self._joinids[axis] is not None
+
+        if column_names is not None:
+            arrow_table = arrow_table.select(column_names)
+        return arrow_table
+
+    def _read_both_axes(
+        self,
+        column_names: AxisColumnNames,
+    ) -> Tuple[pa.Table, pa.Table]:
+        """Private. Read both axes in its entirety, ensure that soma_joinid is retained."""
+        futures = (
+            self._threadpool.submit(
+                self._read_axis_dataframe,
+                "obs",
+                self.experiment.obs,
+                self._matrix_axis_query["obs"],
+                column_names=column_names.get("obs", None),
+            ),
+            self._threadpool.submit(
+                self._read_axis_dataframe,
+                "var",
+                self.experiment.ms[self.measurement_name].var,
+                self._matrix_axis_query["var"],
+                column_names=column_names.get("var", None),
+            ),
+        )
+        concurrent.futures.wait(futures)
+        obs_table, var_table = (f.result() for f in futures)
+        return obs_table, var_table
+
+    def read(
+        self,
+        X_name: str,
+        *,
+        use_position_indexing: bool = False,
+        column_names: Optional[AxisColumnNames] = None,
+        X_layers: Optional[Sequence[str]] = None,
+    ) -> ExperimentAxisQueryReadArrowResult:
+        """
+        Read the _entire_ query result into (in-memory) Arrow Tables. This is a
+        low-level routine intended to be used by loaders for other in-core
+        formats, such as AnnData, which can be created from the resulting Tables.
+        """
+        if column_names is None:
+            column_names = AxisColumnNames(obs=None, var=None)
+        X_collection = self.experiment.ms[self.measurement_name].X
+        X_layers = [] if X_layers is None else X_layers
+        all_X_names = [X_name] + list(X_layers)
+        all_X_arrays: Dict[str, SOMASparseNDArray] = {}
+        for _xname in all_X_names:
+            if not isinstance(_xname, str) or not _xname:
+                raise ValueError("X layer names must be specified as a string.")
+            if _xname not in X_collection:
+                raise ValueError("Unknown X layer name")
+            if not isinstance(X_collection[_xname], SOMASparseNDArray):
+                raise NotImplementedError("Dense array unsupported")
+            all_X_arrays[_xname] = cast(SOMASparseNDArray, X_collection[_xname])
```

Comment:
Because of the `isinstance` check above, I believe this `cast` is no longer necessary.
```suggestion
            all_X_arrays[_xname] = X_collection[_xname]
```

Reply:
mypy isn't smart enough unless I introduce an intermediate variable.  which I will do.